timemachine/md/align.py

```diff
@@ -0,0 +1,96 @@
+import functools
+from typing import Tuple
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from numpy.typing import NDArray
+from rdkit import Chem
+
+from timemachine.fe import topology
+from timemachine.fe.utils import get_mol_masses, get_romol_conf
+from timemachine.ff import Forcefield
+from timemachine.integrator import LangevinIntegrator
+from timemachine.potentials import generic
+
+
+def align_mols_by_core(
+    mol_a: Chem.Mol, mol_b: Chem.Mol, core: NDArray, ff: Forcefield, n_steps: int = 100, k: float = 10000
+) -> Tuple[NDArray, NDArray]:
+    """
+    Given two mols and a core mapping, simulate the mols in vacuum with harmonic bounds between them to get aligned poses.
+
+    Returns the poses with the lowest RMSD between the cores.
+
+    Parameters
+    ----------
+    mol_a: Chem.Mol
+        First Ligand.
+
+    mol_b: Chem.Mol
+        Second Ligand.
+
+    core: np.ndarray
+        K x 2 mapping of coordinates in A to B. Assumes core[:, 0] matches mol_a and core[:, 1] matches mol_b
+
+    ff: ff.Forcefield
+        Wrapper class around a list of handlers.
+
+    n_steps: int
+        Number of steps to run vacuum simulation for.
+
+    k: float
+        Harmonic bond K value
+
+    Returns
+    -------
+    tuple (mol_a_coords, mol_b_coords)
+        New coords of the two molecules.
+
+
+    """
+    mol_a_conf = get_romol_conf(mol_a)
+    mol_b_conf = get_romol_conf(mol_b)
+    x0 = np.concatenate([mol_a_conf, mol_b_conf])
+    masses = np.concatenate([get_mol_masses(mol_a), get_mol_masses(mol_b)])
+    bond_indices = np.array(core)
+    bond_indices[:, 1] += mol_a_conf.shape[0]
+    top = topology.DualTopology(mol_a, mol_b, ff)
+
+    vacuum_system = top.setup_end_state()
+    U = vacuum_system.get_U_fn()
+
+    fb = generic.HarmonicBond(bond_indices)
+    restraint_params = np.zeros((len(core), 2))
+    restraint_params[:, 0] = k
+    bond_U = functools.partial(
+        fb.to_reference(),
+        params=jnp.array(restraint_params),
+        box=None,
+        lam=0.0,
+    )
+
+    def combined_U(x):
+        return U(x) + bond_U(x)
+
+    grad_fn = jax.jit(jax.grad(combined_U, argnums=(0)))
+
+    def force_func(x):
+        du_dx = grad_fn(x)
+        return -du_dx
+
+    # TBD 1e-3 step size is arbitrary
+    intg = LangevinIntegrator(force_func, masses, 0.0, 1e-3, 0.0)
+    key = jax.random.PRNGKey(2022)  # seed value doesn't matter with no friction and 0 temp
+    xs, _ = intg.multiple_steps_lax(key, x0, np.zeros_like(x0), n_steps=n_steps)
+    xs = np.array(xs)
+
+    mol_a_core = xs[:, bond_indices[:, 0]]
+    mol_b_core = xs[:, bond_indices[:, 1]]
+    deltas = mol_a_core - mol_b_core
+    rmsds = [np.linalg.norm(delta) for delta in deltas]
+    best_alignment_idx = np.argmin(rmsds)
```

Comment:
Question: Is taking `idx = argmin(rmsds)` preferable to taking last frame `idx = -1`?

With default restraint force constant k = 10000 kJ/mol / nm^2, I would expect very small restraint RMSDs here (if these are fluctuating @ 300K and there are no other competing energy terms, each restrained atom pair might fluctuate roughly on order of stddev = sqrt(kBT/ k) = 0.015 nm).

Another possibility may be to gradually increase the strength of the core restraint over time and take the last frame, if driving RMSD close to 0 is crucial.

Reply:
Doing `idx = argmin(rmsds)` definitely has an impact. Have seen selecting the last frame produce cores just as bad the current core finding approach we use. It can also explode if the initial core is broken, so this will avoid taking a frame that is outright bad, though could have a separate check outside of `rmsd`. 

> Another possibility may be to gradually increase the strength of the core restraint over time and take the last frame, if driving RMSD close to 0 is crucial.

That certainly seems viable.

Reply:
I am inclined to hold off with the increasing the strength, as it seems like we need to deboggle the integrator and how we configure JAX vacuum systems. Using the GPU systems would work, but this seems too cheap to justify pushing onto the GPU.

Reply:
> as it seems like we need to deboggle the integrator and how we configure JAX vacuum systems

Not sure if this is what needs deboggling, but just a note that now you can convert GPU potentials to JAX reference potentials without too much trouble:

```
from timemachine.potentials import generic

generic.Nonbonded.from_gpu(nonbonded_gpu).to_reference()
```

Reply:
For sure, those would certainly help getting a reference version ready. I like having this a CPU only method and its the JAX potentials and the integrator that would need to be modified to interpolate.

Unless we want to do something like this https://github.com/proteneer/timemachine/blob/master/timemachine/fe/single_topology_v3.py#L656-L679 and change lambda in a loop?

Reply:
Hmm, I think I understand. Could you modify the force function signature to take `k` as an argument, and then step via a for loop while modifying k? Maybe something like

```python
    @jax.jit
    def force_func(x, k):
        restraint_params = jnp.zeros((len(core), 2)).at[:, 0].set(k)

        bond_U = functools.partial(
            U_ref,
            params=jnp.array(restraint_params),
            box=None,
            lam=0.0,
        )

        def combined_U(x):
            return U(x) + bond_U(x)

        dU_dx = jax.grad(combined_U)(x)

        return -du_dx
```

Reply:
That would work, just then need to have a way to modify k as it moves forward. Could recreate a new force func every step along with and integrator, but seems like we should support this more generally rather than hard coding it here.

Reply:
> just then need to have a way to modify k as it moves forward. Could recreate a new force func every step along with and integrator

I'm not sure I understand. In the example the force function takes `k` as a parameter, so it seems like we shouldn't need to create a new force function on every step?